crates/hc_sandbox/src/zome_call.rs

```diff
@@ -0,0 +1,516 @@
+
+    let client = AppClient::try_new(admin_port, zome_call.app_id.clone())
+        .await
+        .context("Could not create app client")?;
+
+    let app_info = client.request(AppRequest::AppInfo).await?;
+    let info = match app_info {
+        AppResponse::AppInfo(Some(info)) => info,
+        other => anyhow::bail!("Unexpected response while getting app info: {:?}", other),
+    };
+
+    let cell_ids = info
+        .cell_info
+        .values()
+        .flatten()
+        .filter_map(|info| match info {
+            CellInfo::Provisioned(info)
+                if info.cell_id.dna_hash() == zome_call.dna_hash.as_ref() =>
+            {
+                Some(info.cell_id.clone())
+            }
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    if cell_ids.is_empty() {
+        anyhow::bail!(
+            "No cell found for DNA hash [{:?}] in app {:?}",
+            zome_call.dna_hash,
+            info
+        );
+    }
+
+    holochain_util::pw::pw_set_piped(zome_call.piped);
+    if !zome_call.piped {
+        msg!("Enter passphrase to authorize zome calls: ");
+    }
+    let passphrase = holochain_util::pw::pw_get().context("Failed to get passphrase")?;
+
+    let (auth, key) = generate_signing_credentials(passphrase)?;
+
+    let (nonce, expires_at) = holochain_nonce::fresh_nonce(Timestamp::now())
+        .map_err(|e| anyhow::anyhow!("Failed to generate nonce: {:?}", e))?;
+
+    let params = ZomeCallParams {
+        provenance: AgentPubKey::from_raw_32(key.verifying_key().as_bytes().to_vec()),
+        cell_id: cell_ids.first().unwrap().clone(),
+        zome_name: ZomeName::from(zome_call.zome_name.clone()),
+        fn_name: FunctionName(zome_call.function.clone()),
+        cap_secret: Some(auth.cap_secret),
+        payload: ExternIO::encode(serde_json::from_slice::<serde_json::Value>(
+            zome_call.payload.as_bytes(),
+        )?)?,
+        nonce,
+        expires_at,
+    };
+    let (payload, hash) = params.serialize_and_hash()?;
+
+    let sig = key.try_sign(&hash)?;
+
+    let response = client
+        .request(AppRequest::CallZome(Box::new(ZomeCallParamsSigned {
+            bytes: ExternIO::from(payload),
+            signature: Signature::try_from(sig.to_vec())?,
+        })))
+        .await?;
+
+    match response {
+        AppResponse::ZomeCalled(response) => {
+            let response: serde_json::Value = response.decode()?;
+            serde_json::to_writer(std::io::stdout(), &response)?;
+            println!(); // Add newline
+        }
+        other => anyhow::bail!("Unexpected response while calling zome: {:?}", other),
+    }
+
+    Ok(())
+}
+
+fn generate_signing_credentials(
+    passphrase: BufRead,
+) -> anyhow::Result<(Auth, ed25519_dalek::SigningKey)> {
+    if unsafe { libsodium_sys::sodium_init() } < 0 {
+        anyhow::bail!("Failed to initialize libsodium");
+    }
+
+    let auth = load_or_create_auth()?;
+
+    let mut out = [0u8; 32];
+    if unsafe {
```

Comment:
It's unfortunate to need `unsafe` code here, but I don't think ed25519-dalek exposes similar functionality.

Reply:
[sodoken exposes this](https://docs.rs/sodoken/0.0.11/sodoken/hash/argon2id/index.html) or [the argon2 crate](https://crates.io/crates/argon2). But also... pwhashing is a very blocking algorithm, it should not be done inline like this, it'll starve tokio tasks.